### PR TITLE
feature: bind user for active_directory auth

### PR DESCRIPTION
**Pull request by murrant**
Optional, allows the use of "remember me", API, and user alerting.

DO NOT DELETE THIS TEXT

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you signed the [Contributors agreement](http://docs.librenms.org/General/Contributing/) - please do NOT submit a pull request unless you have (signing the agreement in the same pull request is fine). Your commit message for signing the agreement must appear as per the docs.
- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`

fixes: #6134

### CODE REVIEW

**Comment by FTBZ** [comment]
I don't know if it's possible, but when the bind user isn't working. It's showing a floating "Invalid credentials" on all pages and allow login of standard AD account (but without access to anything).

A better message should be useful I think.

screen shot 2017-03-24 at 07 38 14

**Comment by FTBZ** [comment]
Noticed just a second annoying thing, the base DN cannot be specified for the bind user? Because our standard users and our technical users like a "bind user" isn't on the same Active Directory OU.

**Comment by murrant** [comment]
Thanks for testing FTBZ, I'll see what I can do.  I think I need an additional check in reauthenticate to make sure the bind user bind was successful.

I'm not sure about the bind user DN,  one way to fix that is to set your base DN higher up the tree.

Maybe, it would make sense to support full DN for the bind user.

**Comment by murrant** [comment]
Actually the bind user doen't use base dn, it uses `${config['auth_ad_binduser']}@${config['auth_ad_domain']}`

**Comment by FTBZ** [comment]
Ref. to #6134 for tracking.

File: doc/Extensions/Authentication.md [hunk 1]
```diff
@@ -41,7 +41,14 @@ When using SELinux on the LibreNMS server, you need to allow Apache (httpd) to c
 setsebool -P httpd_can_connect_ldap=1
 ```
 
-#### MySQL Authentication
+#### Testing authentication
+You can test authentication with this script:
+```shell
+./scripts/auth_test
```
Comment (laf) [inline doc/Extensions/Authentication.md]: scripts/auth_test.php?

**Comment by laf** [comment]
@FTBZ Would you  mind retesting this PR once @murrant is done?

**Comment by murrant** [comment]
All,  done I think I gave it a pretty good round of testing myself.  I would like an additional confirmation from @FTBZ 

I also learned that apparently Firefox does some sort of session caching across browser closes...

**Comment by FTBZ** [comment]
- The floating text message has disappeared and was replaced by the red box on the login screen. It's better, but it's strange to get this kind of message before trying to authenticate. The documentation needs to notified about this if it's not possible to replace the message with the exact error.

screen shot 2017-03-28 at 07 55 15

- I confirm that having two different OU for users and bind user has no impact
- Notification email is working fine, now you can use "Issue alerts to admins" with AD account
- API Access token can be created for AD user now, wasn't possible before without editing the DB

I've a Dashboard problem linked to AD account probably, but the patch returns me a lot of errors when trying to be applied to my production system.

**Comment by murrant** [comment]
Thanks for the update these are all existing issues.

I just made a PR for the dashboard and twofactor bugs. #6286.

The error message can possibly be avoided... might look in to it sometime in a separate PR. It will give verbose output if it is enabled.

**Pull request merged**